File: solutions/elixir/091-min-stack.ex
# Problem 91: Min Stack (LeetCode 155)
# Difficulty: Med
# Language: Elixir
# 
defmodule MinStack do
  def new, do: []

  def push([], x), do: [{x, x}]
  def push([{_, m} | _] = s, x), do: [{x, min(x, m)} | s]

  def pop([_ | rest]), do: rest
  def top([{v, _} | _]), do: v
  def get_min([{_, m} | _]), do: m
end

File: solutions/ocaml/091-min-stack.ml
(*
  Problem 91: Min Stack (LeetCode 155)
  Difficulty: Med
  Language: OCaml
*)
type min_stack = (int * int) list

let push stack x =
  let cur_min = match stack with [] -> x | (_, m) :: _ -> min x m in
  (x, cur_min) :: stack

let pop = function _ :: rest -> rest | [] -> []
let top = function (v, _) :: _ -> v | [] -> failwith "empty"
let get_min = function (_, m) :: _ -> m | [] -> failwith "empty"
